Add new solver and options for torque tube as a spring

eyoung55 opened the pull request:
This PR introduces the ability to simulate the stiffness derived from the torque tube due to fictitious spring forces (Robin boundary conditions). Many changes are included, but in summary:

- Edits to `DomainCreation.py` for the panels3d module to enable
  - a parameter for the separation distance between panel and torque tube center, `pv_array.torque_tube_separation`
  - a parameter for the radius of the torque tube, `pv_array.torque_tube_radius`
  - a parameter for the number of modules in a span/table length, `pv_array.modules_per_span`
  - this defines a mesh with a rectangular standoff about which the rotation of the torque tube is modeled
  - update facet tagging routine to account for the more complicated 3d shape
- Edits to `input_schema.yaml` to allow new parameters with defaults that revert to the old behavior
- Edit to `sim_params.yaml` to slightly reduce the elevation of the 3d panel consistent with defining elevation as the distance between the axis of rotation of the torque tube and the ground level.
- numerous updates to tests to account for the change in meshing conventions, critically, the elevation for cases where the new torque tube features _are not_ specified is now interpreted as the distance between the bottom of the panel and the ground (vs. the centroid of the panel, at the halfway point of thickness, and the ground)
- `ElasticityManager.py` and `StructureMain.py` have updates to save the acceleration data at the NE corner vertex (in addition to the previously saved deformation), rename variables for clarity between deformation and acceleration, and an update to tag the correct vertex with the new mesh convention for elevation

eyoung55 committed:
adding new options for torque tube radius, sep. distance, and modules_per_span

eyoung55 committed:
making panels differently, including torque tube standoff and discrete modules per span

eyoung55 committed:
ignore panel surfaces marked trash, add back in checks on bbox, add new check on z_min/max clipping

eyoung55 committed:
change defaults for torque tube to zero to preserve old mesh behavior by default

eyoung55 committed:
clean up facet marking tests and revert to old meshing behavior if torque tube is not specified (length zero)

eyoung55 committed:
lower elevation by half panel thickness to coincide with new meshing strategy (panel bottom surface = 0 (z))

eyoung55 committed:
change the box to span [0,2] in the z direction vs being 0 centered

eyoung55 committed:
update so that a panel's elevation is defined as the bottom of the panel to the ground

eyoung55 committed:
change dz to account for new panel position, change truth values to account for new positioning conventions

eyoung55 committed:
change the location of the tracked corner acceleration consistent with new meshing convention

eyoung55 committed:
changing variable names from north_west to north_east

eyoung55 committed:
remove misleading comment

eyoung55 committed:
naive parsing to avoid issue of gha node name having underscore

eyoung55 committed:
save both acceleration and deformation in csv file, clean up unclear variable names

eyoung55 committed:
change facets to vertices in variable name for clarity

eyoung55 committed:
only compare the deformation outputs (first two cols) during the check to avoid dimension mismatch

eyoung55 committed:
Merge branch 'dev' into torque-spring

xinhe2205 committed:
update the case study input

Commit:
implement structure simplification

Commit:
dx and ds for modules and connectors

Commit:
new derivations without split the left fixed and right fixed part

Commit:
fix bugs in the input file

Commit:
fix initialization of total torque

xinhe2205 committed:
fix the Assertion Error

The variational form should include only ufl.dx or dx_structure. It cause assertion error if we use ufl.dx and dx_structure at the same time.

xinhe2205 committed:
Initialize spring stiffness

we are solving the structure first. For the first step, no stress is applied to the structure, so the rotation is zero from first time step, which return nan stiffness value. Initialize the spring stiffness to zero at the first time step (or phi=0).

Commit:
verified

Commit:
add lift and drag force calculation

eyoung55 committed:
Merge branch 'dev_wrap' into torque-spring

xinhe2205 committed:
fix facet marker to make it compitile with previous geometry

xinhe2205 committed:
Merge branch 'torque-spring' of https://github.com/eyoung55/PVade into torque-spring

xinhe2205 committed:
fix facet marker for original geometry

xinhe2205 committed:
add back the BCs for case without Robin BC and move the structure down by half panel thickness

xinhe2205 committed:
black format

xinhe2205 commented:
1. For test_solve.py, we cannot reproduce the exactly same mesh as used earlier. Although we can match the geometry (panel_span=21 and all other parameters are the same as those in sim_params.yaml) and fixed lines (3 motor-fixed lines by setting --pv_array.span_fixation_pts=[3.5,10.5,17.5]), the cell number can not match (2513 cells in structure in previous mesh while 2156 cells in new mesh with l_char=20).
 
Looks like the earlier mesh was generated awhile ago, as the center line is broken into 3 segments while it is a single line in the current PVade version (the main repo).

In this case we used the mesh generated by sim_params.yaml to run the test_solve.py, and the truth value are changed.

2. Currently, only the test_mesh_movement.py is not passing from my local test (memory issue).

3. The code are black formatted, but the example folder cannot pass the black formatting check.

arswalid committed:
fix test mesh movement:

- problem was 2 geometries had same volume center
- fix was remove last tag that belongs to fluid
- fix to test file to account for elevation and panel thickness

arswalid committed:
reformatted files in examples

- we should probably exclude those file from black testing and only
  check source files
- black version 26 triggers new formatting errors not triggered by
  version 25

arswalid committed:
Fix to test all input

- input files modified to incorporate panel_ in the naming scheme
- addition of panel_ in domain creation for other geometries
- addition of _0 for top and bottom surfaces when modeling torque tube
  is not available

arswalid committed:
removing -n 4 that was used to test locally back to n 1 for gh runners

arswalid committed:
black formatting

arswalid committed:
fixing misspelled modules/module

arswalid committed:
black formatted latest changes
